crates/jstz_node/src/sequencer/inbox/mod.rs

```diff
@@ -105,15 +102,31 @@ pub async fn spawn_monitor<
     })
 }
 
-/// Process the inbox messages of the given block and push them into the queue.
-async fn process_inbox_messages(
-    block_level: BlockLevel,
+fn stream_factory(endpoint: String) -> impl StreamFactory {
+    move || {
+        let endpoint = endpoint.clone();
+        let stream = async move {
+            api::monitor_blocks(&endpoint)
+                .await
+                .map(|s| s.map_ok(|b| b.level))
+        };
+        futures_util::stream::once(stream).try_flatten().boxed()
+    }
+}
+
+/// Process inbox msgs for the given block:
+/// 1. Filter out irrelevant msgs and parse valid ones into operations.
+/// 2. Push each operation into the shared queue, retrying on failure.
+/// 3. Commit the block as a checkpoint after all operations are queued.
+async fn process_inbox_messages<S: CheckpointStore>(
+    block: &mut PendingBlock<S>,
     block_content: BlockResponse,
     queue: Arc<RwLock<OperationQueue>>,
     ticketer: &ContractKt1Hash,
     jstz: &SmartRollupHash,
 ) {
-    let mut ops = parse_inbox_messages(block_level, block_content, ticketer, jstz);
+    let mut ops = parse_inbox_messages(block.level(), block_content, ticketer, jstz);
+
     while let Some(op) = ops.pop_front() {
         loop {
             let success = queue.write().is_ok_and(|mut q| q.insert_ref(&op).is_ok());
```

Comment:
```suggestion
            let success = queue.write().is_ok_and(|mut q| q.insert(&op).is_ok());
```
insert to prevent unnecessary clone?

Reply:
it's in the loop so it ends up moving